eventsources/sources/amqp/start.go

```diff
@@ -113,50 +116,62 @@ func (el *EventListener) StartListening(ctx context.Context, dispatch func([]byt
 				log.Error("failed to read a message, channel might have been closed")
 				return errors.New("channel might have been closed")
 			}
-			log.Info("received the message", zap.Any("message-id", msg.MessageId))
-			body := &events.AMQPEventData{
-				ContentType:     msg.ContentType,
-				ContentEncoding: msg.ContentEncoding,
-				DeliveryMode:    int(msg.DeliveryMode),
-				Priority:        int(msg.Priority),
-				CorrelationId:   msg.CorrelationId,
-				ReplyTo:         msg.ReplyTo,
-				Expiration:      msg.Expiration,
-				MessageId:       msg.MessageId,
-				Timestamp:       msg.Timestamp.String(),
-				Type:            msg.Type,
-				AppId:           msg.AppId,
-				Exchange:        msg.Exchange,
-				RoutingKey:      msg.RoutingKey,
-				Metadata:        amqpEventSource.Metadata,
-			}
-			if amqpEventSource.JSONBody {
-				body.Body = (*json.RawMessage)(&msg.Body)
-			} else {
-				body.Body = msg.Body
-			}
-
-			bodyBytes, err := json.Marshal(body)
-			if err != nil {
-				log.Error("failed to marshal the message", zap.Any("message-id", msg.MessageId), zap.Error(err))
-				continue
-			}
-
-			log.Info("dispatching event ...")
-			err = dispatch(bodyBytes)
-			if err != nil {
-				log.Error("failed to dispatch AMQP event", zap.Error(err))
+			if err := el.handleOne(amqpEventSource, msg, dispatch, log); err != nil {
+				log.Errorw("failed to process an AMQP message", zap.Error(err))
+				el.Metrics.EventProcessingFailed(el.GetEventSourceName(), el.GetEventName())
 			}
 		case <-ctx.Done():
 			err = conn.Close()
 			if err != nil {
-				log.Error("failed to close connection", zap.Error(err))
+				log.Errorw("failed to close connection", zap.Error(err))
 			}
 			return nil
 		}
 	}
 }
 
+func (el *EventListener) handleOne(amqpEventSource *v1alpha1.AMQPEventSource, msg amqplib.Delivery, dispatch func([]byte) error, log *zap.SugaredLogger) error {
+	startTime := time.Now()
+	defer func(start time.Time) {
```

Comment:
Do you mean something like following? No it's wrong.

```
defer el.Metrics.EventProcessingDuration(el.GetEventSourceName(), el.GetEventName(), float64(time.Since(time.Now())/time.Millisecond))
```

But it can be simplified as 

```
defer func(start time.Time) {
  el.Metrics.EventProcessingDuration(el.GetEventSourceName(), el.GetEventName(), float64(time.Since(start)/time.Millisecond))
}(time.Now())
```